File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        
        maxDep = 1
        left = self.maxDepth(root.left)
        right = self.maxDepth(root.right)
        
        return max(left, right) +1
        
#         time: o(n), space: o(1)
#         if not root: return 0
        
#         cur = 0
#         que = [root]
        
#         while que:
#             for i in range(len(que)):
#                 curNode = que.pop(0)

#                 if curNode.left:
#                     que.append(curNode.left)
#                 if curNode.right:
#                     que.append(curNode.right)
#             cur += 1

    
#         return cur
    
        

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    /*
    dfs
    1. check a null case
    2. make a func to count depth
    3. get the max depth to compare 
    */
    
    if(!root) return 0;
    let left = maxDepth(root.left);
    let right = maxDepth(root.right);
    
    return Math.max(left, right) + 1
    
};
